From 333c3082b08ad5ee44219dabbf4e4cc919136ffd Mon Sep 17 00:00:00 2001
From: Chuanxiao Dong <chuanxiao.dong@intel.com>
Date: Tue, 27 Nov 2012 16:19:03 +0800
Subject: [PORT FROM MAIN] vold: change vold so that vold can skip the scan on extend partition

BZ: 47043

Below is the original commit info
	BZ: 17252

	Using the tool fsck_msdos to check extend partition will cause libc error.
	Change vold so that vold can skip such partition.

Change-Id: I10c47f13a318ce1b838a965bc3767c90d2ed27f3
Orig-Change-Id: I9d4832461e39e061f130115ac06fe4e843d85afb
Category: aosp improvement
Domain: Storage Removable_Storage
Origin: internal
Upstream-Candidate: no, proprietary
Signed-off-by: Chuanxiao Dong <chuanxiao.dong@intel.com>
Reviewed-on: http://android.intel.com:8080/76535
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Tested-by: Sun, Jianhua <jianhua.sun@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>

Conflicts:
	Fat.cpp
	Fat.h
---
 DirectVolume.h |    4 +---
 Fat.cpp        |   35 +++++++++++++++++++++++++++++++++++
 Fat.h          |    1 +
 Volume.cpp     |   22 ++++++++++++++++++----
 Volume.h       |    3 +++
 5 files changed, 58 insertions(+), 7 deletions(-)

diff --git a/DirectVolume.h b/DirectVolume.h
index 7be133f..ba917a5 100644
--- a/DirectVolume.h
+++ b/DirectVolume.h
@@ -21,13 +21,11 @@
 
 #include "Volume.h"
 
-#define MAX_PARTS 4
-
 typedef android::List<char *> PathCollection;
 
 class DirectVolume : public Volume {
 public:
-    static const int MAX_PARTITIONS = 32;
+    static const int MAX_PARTITIONS = MAX_PARTS;
 protected:
     const char* mMountpoint;
     const char* mFuseMountpoint;
diff --git a/Fat.cpp b/Fat.cpp
index c967a90..8206f9f 100644
--- a/Fat.cpp
+++ b/Fat.cpp
@@ -253,3 +253,38 @@ void Fat::wipe(const char *fsPath, unsigned int numSectors) {
         SLOGE("Fat wipe failed to open device %s", fsPath);
     }
 }
+
+int Fat::check_extend(const char *fsPath, unsigned int numParts)
+{
+    int n = numParts, i, fd;
+    unsigned char buf[512];
+    unsigned char *PartInfo;
+    int PartInfoOffset = 0x1be;
+    int sizeofPart = 16;
+
+    fd = open(fsPath, O_RDONLY);
+    if (fd < 0) {
+         /* Badness - abort the mount */
+         SLOGE("%s failed FS open (%s)", fsPath, strerror(errno));
+         return -1;
+    }
+
+    if (read(fd, buf, 512) < 0) {
+        /* Badness - abort the mount */
+        SLOGE("%s failed MBR read (%s)", fsPath, strerror(errno));
+        close(fd);
+        return -1;
+    }
+    close(fd);
+
+    PartInfo = buf + PartInfoOffset;
+
+    for (i = 0; i < n; i++, PartInfo += sizeofPart) {
+        if (PartInfo[4] == 5) {
+            SLOGW("%s This is an Extend Partition, skip to next one, n is %d\n", fsPath, i);
+            return i;
+        }
+    }
+
+    return -1;
+}
diff --git a/Fat.h b/Fat.h
index 19614d1..77ac3dd 100644
--- a/Fat.h
+++ b/Fat.h
@@ -26,6 +26,7 @@ public:
                        bool ro, bool remount, bool executable,
                        int ownerUid, int ownerGid, int permMask,
                        bool createLost);
+    static int check_extend(const char *fsPath, unsigned int numParts);
     static int format(const char *fsPath, unsigned int numSectors, bool wipe);
 
 private:
diff --git a/Volume.cpp b/Volume.cpp
index 7c91d8b..1ab3017 100644
--- a/Volume.cpp
+++ b/Volume.cpp
@@ -248,7 +248,7 @@ int Volume::formatVol(bool wipe) {
     }
 
     bool formatEntireDevice = (mPartIdx == -1);
-    char devicePath[255];
+    char devicePath[PATH_MAX];
     dev_t diskNode = getDiskDevice();
     dev_t partNode = MKDEV(MAJOR(diskNode), (formatEntireDevice ? MINOR(diskNode) : MINOR(diskNode) + mPartIdx));
 
@@ -322,7 +322,8 @@ bool Volume::isMountpointMounted(const char *path) {
 }
 
 int Volume::mountVol() {
-    dev_t deviceNodes[4];
+    dev_t deviceNodes[MAX_PARTS];
+    dev_t diskdevice;
     int n, i, rc = 0;
     char errmsg[255];
 
@@ -335,6 +336,8 @@ int Volume::mountVol() {
     char crypto_state[PROPERTY_VALUE_MAX];
     char encrypt_progress[PROPERTY_VALUE_MAX];
     int nParts = 0;
+    char devicePath[PATH_MAX];
+    int extendPart = -1;
 
     property_get("vold.decrypt", decrypt_state, "");
     property_get("vold.encrypt_progress", encrypt_progress, "");
@@ -367,7 +370,7 @@ int Volume::mountVol() {
         return 0;
     }
 
-    nParts = getDeviceNodes((dev_t *) &deviceNodes, 4);
+    nParts = getDeviceNodes((dev_t *) &deviceNodes, MAX_PARTS);
     if (!nParts)
         n = 1;
     else
@@ -422,8 +425,19 @@ int Volume::mountVol() {
         }
     }
 
+    /*
+     * Since we support multiple partitions, in case there is extend SD card
+     * partition, we need to skip this extend partition
+     */
+    diskdevice = getDiskDevice();
+    sprintf(devicePath, "/dev/block/vold/%d:%d", MAJOR(diskdevice),
+        MINOR(diskdevice));
+    extendPart = Fat::check_extend(devicePath, n);
+
     for (i = 0; i < n; i++) {
-        char devicePath[255];
+
+        if (i == extendPart)
+            continue;
 
         sprintf(devicePath, "/dev/block/vold/%d:%d", MAJOR(deviceNodes[i]),
                 MINOR(deviceNodes[i]));
diff --git a/Volume.h b/Volume.h
index 1444ed3..90c4bc6 100644
--- a/Volume.h
+++ b/Volume.h
@@ -20,6 +20,9 @@
 #include <utils/List.h>
 #include <fs_mgr.h>
 
+#define MAX_PARTS 32
+#define PATH_MAX  255
+
 class NetlinkEvent;
 class VolumeManager;
 
-- 
1.7.4.1

